crates/bevy_math/src/compass.rs

```diff
@@ -0,0 +1,397 @@
+use crate::Dir2;
+#[cfg(feature = "bevy_reflect")]
+use bevy_reflect::Reflect;
+#[cfg(all(feature = "serialize", feature = "bevy_reflect"))]
+use bevy_reflect::{ReflectDeserialize, ReflectSerialize};
+
+/// A compass enum with 4 directions.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Debug, PartialEq))]
+#[cfg_attr(
+    all(feature = "serialize", feature = "bevy_reflect"),
+    reflect(Deserialize, Serialize)
+)]
+pub enum CompassQuadrant {
+    /// Corresponds to [`Dir2::Y`] and [`Dir2::NORTH`]
+    North,
+    /// Corresponds to [`Dir2::X`] and [`Dir2::EAST`]
+    East,
+    /// Corresponds to [`Dir2::NEG_X`] and [`Dir2::SOUTH`]
+    South,
+    /// Corresponds to [`Dir2::NEG_Y`] and [`Dir2::WEST`]
+    West,
+}
+
+/// A compass enum with 8 directions.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Debug, PartialEq))]
+#[cfg_attr(
+    all(feature = "serialize", feature = "bevy_reflect"),
+    reflect(Deserialize, Serialize)
+)]
+pub enum CompassOctant {
+    /// Corresponds to [`Dir2::Y`] and [`Dir2::NORTH`]
+    North,
+    /// Corresponds to [`Dir2::NORTH_EAST`]
+    NorthEast,
+    /// Corresponds to [`Dir2::X`] and [`Dir2::EAST`]
+    East,
+    /// Corresponds to [`Dir2::SOUTH_EAST`]
+    SouthEast,
+    /// Corresponds to [`Dir2::NEG_X`] and [`Dir2::SOUTH`]
+    South,
+    /// Corresponds to [`Dir2::SOUTH_WEST`]
+    SouthWest,
+    /// Corresponds to [`Dir2::NEG_Y`] and [`Dir2::WEST`]
+    West,
+    /// Corresponds to [`Dir2::NORTH_WEST`]
+    NorthWest,
+}
+
+impl From<CompassQuadrant> for Dir2 {
+    fn from(q: CompassQuadrant) -> Self {
+        match q {
+            CompassQuadrant::North => Dir2::NORTH,
+            CompassQuadrant::East => Dir2::EAST,
+            CompassQuadrant::South => Dir2::SOUTH,
+            CompassQuadrant::West => Dir2::WEST,
+        }
+    }
+}
+
+impl From<Dir2> for CompassQuadrant {
+    fn from(dir: Dir2) -> Self {
+        let angle = dir.to_angle().to_degrees();
+
+        match angle {
+            -135.0..=-45.0 => Self::South,
+            -45.0..=45.0 => Self::East,
+            45.0..=135.0 => Self::North,
+            135.0..=180.0 | -180.0..=-135.0 => Self::West,
+            _ => unreachable!(),
+        }
+    }
+}
```

Comment:
I'm not sure I agree with this implementation. It's too easy to convert from a `Dir2` to a `CompassQuadrant`, but it does not explicitly document that this is a lossy conversion. For instance:

```rust
let original_angle = 13.0;
let dir2 = Dir2::new(Vec2::from_angle(original_angle)).unwrap();
let compass = CompassQuadrant::from(dir2);
let dir2_lossy = Dir2:from(compass);

// Panics, because the original angle is lost.
assert_eq!(dir2_lossy.to_angle().to_degrees(), original_angle);
```

People tend to call `.into()` without reading about the consequences, just to fix their code. I would much prefer this be converted into an explicit method that you have to opt-in to using, so you are aware of the consequences.

Reply:
I wouldn't expect a lossless conversion? Maybe it should be documented, but expecting to get an original angle back seems...odd?

Reply:
I worry that people will miss the fact that the conversion is lossy because `.into()` is not explicit about the resulting type without rust-analyzer.

```rust
let foo = dir2.into();
```

If `foo` wasn't named something like `compass_quadrant`, you wouldn't know that it's converted to a `CompassQuadrant`, and thus wouldn't know that data is being lost.

At the very least, could you add documentation to the `from()` method that says the conversion is lossy? The problem may not be large enough to deem a separate method, but it should be apparent in more than just the source code.

Reply:
Done